Mail: include markdown in mails

lukastocker opened the pull request:
- add markdown rendering. Also always escape HTML in markdown output
- use UI form in ilMailFormGUI and add Markdown for Message
- enable markdown for free text mails
- include new functions in form and buttons to avoid a redirect of the page if the user clicked onto the placeholders to insert them into the mail
- adjust template rendering in 'withAdditionalOnLoadCode' if coming from a context
- adjust PHPUnit Tests
- use ResourceCollection for uploading/showing file information

lukastocker commented:
This PR replaces Draft [7779](https://github.com/ILIAS-eLearning/ILIAS/pull/7779) which is now closed.

It was decided with @mjansenDatabay that we will use the ResourceCollection for storing, converting and displaying the file data instead of rebuilding everything to IRSS. The needed functions can be found in a trait file as they are needed multiple times.

The placeholders underneath the message field are always visible as there's currently no UI alternative to the "old" implementation. 

The search for recipients (in the tag input fields to, cc, bcc) still might take a long time to load as it was decided that we wait for an server source extension from Stefan K. 

In addition I created three more PR's you could check out:

1. [Underline Extension for Markdown](https://github.com/ILIAS-eLearning/ILIAS/pull/9425)
2. Adjusting the [Markdown's Headline configuration](https://github.com/ILIAS-eLearning/ILIAS/pull/9424), which is also needed for the underline extension PR
3. [Optional Markdown Extension with emphasis rendering](https://github.com/ILIAS-eLearning/ILIAS/pull/9683)

klees commented:
Hi @mjansenDatabay,

could you have another look at this? Feel free to ping @lukastocker for a rebase if we are approaching a merge.

Kind regards!

klees commented:
@lukastocker Could you please:

* Remove the markdown extension that we have in the other PRs. They should be independent from this...
* Open a single PR for the changes in the UI framework (but still leave the changes in here...)

Ping me afterwards please =)

Thanks!

mjansenDatabay commented:
Hi @klees ,

I will have a look (end of) next week.

Best regards,
Michael

lukastocker commented:
> @lukastocker Could you please:
> 
> * Remove the markdown extension that we have in the other PRs. They should be independent from this...
> * Open a single PR for the changes in the UI framework (but still leave the changes in here...)
> 
> Ping me afterwards please =)
> 
> Thanks!

Hi @klees ,

i just updated this PR (removal of markdown extension) and added following new PR's:

1. [Optional Extension for markdown emphasis rendering](https://github.com/ILIAS-eLearning/ILIAS/pull/9683)
2. [Additional PR for markdown UI changes](https://github.com/ILIAS-eLearning/ILIAS/pull/9684)

lukastocker committed:
Markdown: adjust attachment handling and adjust pull conflicts with release_11.

mjansenDatabay commented:
I will be next week, sorry ...

And: Thanks for your changes, Luka.

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,109 @@
+        return $collection;
+    }
+
+    public function FilesFromLegacyToIRSS(array $mail_data): array
```
Please provide the shape of the arrays as PHPDoc comments.

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,109 @@
+trait FileDataRCHandling
+{
+    protected function getCurrentCollection(
+        array $path_to_files,
```
Please provide the shape of the array as PHPDoc comments.

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,109 @@
+        return $files;
+    }
+
+    public function getIDforCollection(array $mail_data): ?ResourceCollectionIdentification
```
Please provide the shape of the array as PHPDoc comments.

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,109 @@
+        return $rcid;
+    }
+
+    public function FilesFromIRSSToLegacy(ResourceCollectionIdentification $identification): array
```
Please provide the shape of the arrays as PHPDoc comments.

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,109 @@
+        return $files;
+    }
+
+    protected function handleAttachments(array $attachments): array
```
Please provide the shape of the arrays as PHPDoc comments.

mjansenDatabay commented on components/ILIAS/Init/classes/class.ilInitialisation.php:
```diff
@@ -715,6 +715,27 @@ protected static function setSessionCookieParams(): void
         }
     }
 
+    protected static function initMail(\ILIAS\DI\Container $c): void
```
Can you try to move this to the `Mail.php` in the component folder, and if it is not possible (due to effort, costs, etc.), please move this to  `\ILIAS\Mail\Service\MailService`.

I also like explicit types everywhere (if possible), so please provide return type declarations for the functions.

mjansenDatabay commented on components/ILIAS/Mail/classes/Preview/class.ilMailPreviewGUI.php:
```diff
@@ -0,0 +1,83 @@
+
+class ilMailPreviewGUI
+{
+    protected ilMailTemplate $template_id;
+    protected ilMailTemplatePlaceholderResolver $placeholder_resolver;
+    protected ILIAS\Refinery\Factory $refinery;
+    protected ilGlobalTemplateInterface $tpl;
```
We can make them `private readonly` IMO.

lukastocker replied:
As we do not use this class I removed it. It was an old implementation part from my side.

mjansenDatabay commented on components/ILIAS/Mail/classes/Preview/ilPreviewFactory.php:
```diff
@@ -0,0 +1,32 @@
+
+declare(strict_types=1);
+
+class ilPreviewFactory
```
As far as I can see we don't use this class at all.

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,109 @@
+    {
+        $files = [];
+        $path_to_files = [];
+        $attachment = unserialize($mail_data['attachments']);
```
Can you please add the `allowed_classes` option to address security concerns here?

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,109 @@
+        $rcid = $this->storage->collection()->id();
+        $collection = $this->storage->collection()->get($rcid);
+        foreach ($path_to_files as $path_to_file) {
+            $resource = @fopen($path_to_file, 'rb');
```
We should try to avoid the `@` operator. This was a common agreement we all made during the "PHP 8 Project".

A pattern/an approach to avoid it is to use `try/cach/finally`, set a custom error handler like ...

```php
set_error_handler(static function (int $severity, string $message, string $file, int $line): void {
    throw new ErrorException($message, $severity, $severity, $file, $line);
});
```
..., and reset the error handler in the `finally` block by calling `restore_error_handler();`.

mjansenDatabay commented on components/ILIAS/Mail/classes/BackgroundTask/class.ilMailDeliveryJob.php:
```diff
@@ -50,29 +50,20 @@ public function run(array $input, Observer $observer): Value
             $mail = new ilFormatMail((int) $input[0]->getValue());
         }
 
-        if (isset($context_parameters['auto_responder'])) {
```
We must not remove this feature here.

I don't think we need to change this class, maybe it was changed accidentally.

mjansenDatabay commented on components/ILIAS/Mail/classes/BackgroundTask/class.ilMassMailDeliveryJob.php:
```diff
@@ -29,8 +29,8 @@
 {
-    private readonly ILIAS\DI\Container $dic;
-    private readonly ilMailValueObjectJsonService $mail_json_service;
+    private ILIAS\DI\Container $dic;
```
I don't think we need to change this class, maybe it was changed accidentally.

mjansenDatabay commented on components/ILIAS/Mail/classes/BackgroundTask/class.ilMailDeliveryJob.php:
```diff
@@ -50,29 +50,20 @@ public function run(array $input, Observer $observer): Value
             ->withContextParameters($context_parameters);
 
-        $mail_data = new MailDeliveryData(
+        $mail->sendMail(
```
Why do you suggest getting rid of the `MailDeliveryData` variable and passing the parameters as primitive types?

lukastocker replied:
I reset this changes into the "old" structure

mjansenDatabay commented on components/ILIAS/Course/classes/class.ilObjCourseGUI.php:
```diff
@@ -214,9 +214,12 @@ public function infoScreen(): void
-            ));
+            $info->addProperty(
+                $this->lng->txt('crs_syllabus'),
+                $this->refinery->string()->markdown()->toHTML()->transform(
```
Do we really need the `markdown` transformation here? Can't we use `$this->refinery->string()->makeClickable(...)`?

lukastocker replied:
We can do that for sure

mjansenDatabay commented on components/ILIAS/Mail/classes/BackgroundTask/class.ilMassMailTaskProcessor.php:
```diff
@@ -29,16 +29,17 @@ class ilMassMailTaskProcessor
     private readonly ilLanguage $language;
     private readonly ilLogger $logger;
-    private readonly ilMailValueObjectJsonService $object_json_service;
+    private readonly ilMailValueObjectJsonService $objectJsonService;
```
I don't think we need to change this class, maybe it was changed accidentally.

mjansenDatabay commented on components/ILIAS/Mail/classes/Mime/Transport/class.ilMailMimeTransportBase.php:
```diff
@@ -88,11 +88,19 @@ final public function send(ilMimeMail $mail): bool
+            )) {
+                ilLoggerFactory::getLogger('mail')->warning($this->getMailer()->ErrorInfo);
+            }
+        }
```
This looks like a duplicate code block.

mjansenDatabay commented on components/ILIAS/Mail/classes/Mime/Transport/class.ilMailMimeTransportBase.php:
```diff
@@ -88,11 +88,19 @@ final public function send(ilMimeMail $mail): bool
-        )) {
+                $mail->getFrom()->getReplyToAddress(),
+                $mail->getFrom()->getReplyToName()
+            )) {
```
Please leave the formatting as it is.

mjansenDatabay commented on components/ILIAS/Mail/classes/Mime/Sender/class.ilMailMimeSenderUser.php:
```diff
@@ -75,19 +75,13 @@ public function getFromName(): string
-
-        if ($template !== $interpolated) {
-            return $interpolated;
+        $name = str_ireplace('[FULLNAME]', $this->user->getFullname(), $from);
```
I don't think we need to change this class, maybe it was changed accidentally.

mjansenDatabay commented on components/ILIAS/Mail/classes/Object/class.ilMailValueObject.php:
```diff
@@ -20,28 +20,37 @@
 {
-    /** @var list<string> */
-    private readonly array $attachments;
+    /** @var string[] */
```
I don't think we need to change this class, maybe it was changed accidentally.

mjansenDatabay commented on components/ILIAS/Mail/classes/Preview/class.ilMailPreviewGUI.php:
```diff
@@ -0,0 +1,83 @@
+    {
+        return $this->placeholder_resolver->resolveForPreview($message);
+    }
+}
```
And empty line is missing at the end of the file (code style).

mjansenDatabay commented on components/ILIAS/Mail/classes/Setup/Migration/class.ilUpdateMailTemplatesForMustache.php:
```diff
@@ -0,0 +1,104 @@
+use ILIAS\Setup\Migration;
+use ILIAS\Setup\Environment;
+
+class ilUpdateMailTemplatesForMustache implements Migration
```
I don't think we need to reintroduce this class, maybe it was added accidentally.

mjansenDatabay commented on components/ILIAS/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php:
```diff
@@ -290,4 +290,28 @@ public function step_15(): void
+
+    public function step_17(): void
+    {
+        if ($this->db->tableColumnExists("mail_saved", "attachments")) {
```
```suggestion
        if ($this->db->tableColumnExists('mail_saved', 'attachments')) {
```

mjansenDatabay commented on components/ILIAS/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php:
```diff
@@ -290,4 +290,28 @@ public function step_15(): void
+
+    public function step_16(): void
+    {
+        if ($this->db->tableExists("mail_saved")) {
```
```suggestion
        if ($this->db->tableExists('mail_saved')) {
```

mjansenDatabay commented on components/ILIAS/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php:
```diff
@@ -290,4 +290,28 @@ public function step_15(): void
+    public function step_16(): void
+    {
+        if ($this->db->tableExists("mail_saved")) {
+            $query = "UPDATE mail_saved SET " . PHP_EOL
```
According to our discussion some time ago, we should set the value to `null` if possible. As far as I can see, the field is nullable.

> Via DBUpdateObjective im setup das Feld "attachment" für alle existierenden DB-Records auf NULL setzen

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilAccountMail.php:
```diff
@@ -44,7 +44,7 @@ public function __construct()
         $this->settings = $DIC->settings();
         $this->repository_tree = $DIC->repositoryTree();
-        $this->sender_factory = $DIC->mail()->mime()->senderFactory();
+        $this->sender_factory = $DIC['mail.mime.sender.factory'];
```
I don't think we need to change this class, maybe it was changed accidentally.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilFileDataMail.php:
```diff
@@ -145,7 +145,10 @@ public function getAttachmentPath(string $a_filename, int $a_mail_id): string
 
     /**
-     * @param list<string> $a_attachments
+     * Adopt attachments (in case of forwarding a mail)
```
I don't think we need to change this class, maybe it was changed accidentally.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailAttachmentTableGUI.php:
```diff
@@ -0,0 +1,99 @@
+<?php
```
I don't think we need to reintroduce this class, maybe it was added accidentally.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailAttachmentGUI.php:
```diff
@@ -29,17 +29,23 @@
     private readonly ilTabsGUI $tabs;
     private AttachmentManagement $mode = AttachmentManagement::MANAGE;
 
+    protected ILIAS\ResourceStorage\Services $storage;
```
```suggestion
    private readonly ILIAS\ResourceStorage\Services $storage;
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailAttachmentGUI.php:
```diff
@@ -29,17 +29,23 @@
 
+    protected ILIAS\ResourceStorage\Services $storage;
+
+    protected ilMailFormUploadHandlerGUI $upload_handler;
```
```suggestion
    private readonly ilMailFormUploadHandlerGUI $upload_handler;
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailTemplateGUI.php:
```diff
@@ -582,4 +582,4 @@ public function setAsContextDefault(): void
         $this->ctrl->redirect($this, 'showTemplates');
     }
-}
+}
```
I don't think we need to change this class, maybe it was changed accidentally.

mjansenDatabay commented on components/ILIAS/Mail/tests/ilMailMimeTest.php:
```diff
@@ -86,7 +92,7 @@ public function testTransportFactoryWillReturnNullTransportIfExternalEmailDelive
         ])->getMock();
         $settings->method('get')->willReturnCallback(static function ($key): ?string {
-            return (string) ('mail_allow_external' !== $key);
+            return (string) !('mail_allow_external' === $key);
```
I suggest leaving this as it is.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailTemplatePlaceholderResolver.php:
```diff
@@ -31,15 +31,25 @@ public function resolve(
                 $context_parameters,
-                $user
+                $user,
+                $replace_empty_placeholders
```
Do we really need this change? The `$replae_empty_placeholders` parameter was removed with https://github.com/ILIAS-eLearning/ILIAS/commit/95cf20df24925a5856a876d77c203274d73b5794#diff-5d6cc6aba611f92504a8557048c6223f7540fe1145fdd2c3b2c18fbf26ce7aaf by your colleague Stefan.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailTemplateContext.php:
```diff
@@ -116,7 +116,8 @@ abstract public function getSpecificPlaceholders(): array;
         array $context_parameters,
-        ?ilObjUser $recipient = null
+        ?ilObjUser $recipient = null,
+        bool $html_markup = false
```
Are you sure we need this? Your colleague Stefan removed this `$html_markup` with https://github.com/ILIAS-eLearning/ILIAS/commit/fbe4f34cb5780d70f22efd05e948f953da53693c#diff-1bc313c6b9ce167839e7a7eeb1c75e34b9aec217f340721a0345c746fb568662

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailTemplateContext.php:
```diff
@@ -125,7 +126,8 @@ abstract public function resolveSpecificPlaceholder(
         array $context_parameters,
-        ?ilObjUser $recipient = null
+        ?ilObjUser $recipient = null,
+        bool $html_markup = false
```
Are you sure we need this? Your colleague Stefan removed this `$html_markup` parameter with https://github.com/ILIAS-eLearning/ILIAS/commit/fbe4f34cb5780d70f22efd05e948f953da53693c#diff-1bc313c6b9ce167839e7a7eeb1c75e34b9aec217f340721a0345c746fb568662

mjansenDatabay commented on components/ILIAS/Mail/tests/ilMailTaskProcessorTest.php:
```diff
@@ -92,13 +92,13 @@ public function testOneTask(): void
             $this->logger_mock,
             $this->dic_mock,
-            new ilMailValueObjectJsonService()
+            new ilMailValueObjectJsonService(),
```
I don't think we need to change this class, maybe it was changed accidentally.

mjansenDatabay commented on components/ILIAS/Mail/tests/ilMailTest.php:
```diff
@@ -29,11 +29,16 @@
-    private MockObject&ilLogger $mock_log;
-    private MockObject&ilMailRfc822AddressParserFactory $mock_parser_factory;
-    private MockObject&ilLanguage $mock_language;
+    /** @var MockObject&ilDBInterface */
```
Please only change the relevant code in this class, so:

- Please keep the code/formatting style as it is.
- Please don't remove the "Intersection Types".
- Please don't rename variables (use the current state of the `trunk` branch).
- Please don't change `$this` to `self` if not necessary, please keep the state of the current `trunk`.

mjansenDatabay commented on components/ILIAS/Mail/tests/ilMailTest.php:
```diff
@@ -601,7 +608,9 @@ private function create(int $ref_id = 234, int $usr_id = 123): ilMail
+        return $instance;
     }
-}
+}
```
The file/line ending violates our code style.

mjansenDatabay commented on components/ILIAS/Mail/tests/ilMailTemplateContextTest.php:
```diff
@@ -57,7 +57,8 @@ public function getSpecificPlaceholders(): array
                 string $placeholder_id,
                 array $context_parameters,
-                ?ilObjUser $recipient = null
+                ?ilObjUser $recipient = null,
```
This is probably not necessary, depending on your answer regarding my remarks in the code to test (class `ilMailTemplateContext`).

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailTemplateGenericContext.php:
```diff
@@ -43,7 +43,8 @@ public function getSpecificPlaceholders(): array
         array $context_parameters,
-        ?ilObjUser $recipient = null
+        ?ilObjUser $recipient = null,
+        bool $html_markup = false
```
Do we really need this change? The `$replae_empty_placeholders` parameter was removed with https://github.com/ILIAS-eLearning/ILIAS/commit/fbe4f34cb5780d70f22efd05e948f953da53693c#diff-4a3f242ad318bf7605be44cf91bf1dee391bde276ea42e5ed5b8649e8831c47a by your colleague Stefan.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailTemplateContext.php:
```diff
@@ -55,7 +55,7 @@ abstract public function getDescription(): string;
      * @return array{mail_salutation: array{placeholder: string, label: string}, first_name: array{placeholder: string, label: string}, last_name: array{placeholder: string, label: string}, login: array{placeholder: string, label: string}, title: array{placeholder: string, label: string, supportsCondition: true}, firstname_lastname_superior: array{placeholder: string, label: string}, ilias_url: a ...
      */
-    private function getGenericPlaceholders(): array
+    protected function getGenericPlaceholders(): array
```
Can we leave this `private`?

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormUploadHandlerGUI.php:
```diff
@@ -0,0 +1,127 @@
+
+/**
+ * Class ilMailFormUploadHandlerGUI
+ */
```
```suggestion
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormUploadHandlerGUI.php:
```diff
@@ -0,0 +1,127 @@
+    {
+        $this->upload->process();
+        /**
+         * @var $result UploadResult
```
```suggestion
         * @var UploadResult $result
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormUploadHandlerGUI.php:
```diff
@@ -0,0 +1,127 @@
+            $i = $this->storage->manage()->upload($result, $this->stakeholder);
+            $status = HandlerResultInterface::STATUS_OK;
+            $identifier = $i->serialize();
+            $message = 'Upload ok';
```
Should't we use a language variable for this?

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormUploadHandlerGUI.php:
```diff
@@ -0,0 +1,127 @@
+    /**
+     * @var ilMailAttachmentStakeholder
+     */
+    private $stakeholder;
```
Can we use typed properties and the `reaonly` modifier?

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailAttachmentGUI.php:
```diff
@@ -29,17 +29,23 @@
 
+    protected ILIAS\ResourceStorage\Services $storage;
+
+    protected ilMailFormUploadHandlerGUI $upload_handler;
```
Can you elaborate, why need this property here? `ilMailAttachmentGUI` already extends `AbstractCtrlAwareUploadHandler`, which is also valid for `ilMailFormUploadHandlerGUI`. So can you explain the purpose of the latter?

lukastocker replied:
That seemed to be some leftovers and I just didn't remove them after overthrowing my old solution. Thanks for the hint!

mjansenDatabay commented on components/ILIAS/Mail/classes/Mime/Transport/class.ilMailMimeTransportBase.php:
```diff
@@ -113,7 +121,7 @@ final public function send(ilMimeMail $mail): bool
         }
 
-        if ($mail->getFinalBodyalt() !== '') {
+        if ($mail->getFinalBodyAlt()) {
```
We should leave this as it is.

mjansenDatabay commented on components/ILIAS/Mail/classes/Mime/Transport/class.ilMailMimeTransportBase.php:
```diff
@@ -181,9 +189,14 @@ final public function send(ilMimeMail $mail): bool
         $this->event_handler->raise('components/ILIAS/Mail', 'externalEmailDelegated', [
             'mail' => $mail,
-            'result' => $result,
+            'result' => $result
```
We should leave this as it is.

mjansenDatabay commented on components/ILIAS/Mail/classes/Mime/Transport/class.ilMailMimeTransportBase.php:
```diff
@@ -181,9 +189,14 @@ final public function send(ilMimeMail $mail): bool
+    public function getErrorInformation(): string
+    {
+        return $this->getMailer()->ErrorInfo;
+    }
```
This method is not used, so we should remove it.

mjansenDatabay commented on components/ILIAS/Mail/classes/Preview/class.ilMailPreviewGUI.php:
```diff
@@ -0,0 +1,83 @@
+
+declare(strict_types=1);
+
+class ilMailPreviewGUI
```
This class seems not to be used anywhere.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailTemplatePlaceholderResolver.php:
```diff
@@ -31,15 +31,25 @@ public function resolve(
+            $message,
+            new ilMailPreviewContextAdapter()
+        );
+    }
```
This method is not used because the consuming class `ilMailPreviewGUI` is not used.

mjansenDatabay commented on components/ILIAS/Mail/classes/Mustache/class.ilMailPreviewContextAdapter.php:
```diff
@@ -0,0 +1,32 @@
+
+declare(strict_types=1);
+
+class ilMailPreviewContextAdapter
```
This class is not used, because `\ilMailTemplatePlaceholderResolver::resolveForPreview` is only consumed by `ilMailPreviewGUI`, which is not used at all.

lukastocker replied:
Additionally I removed the function mentioned by you in ilMailTemplatePlaceholderResolver

mjansenDatabay commented on components/ILIAS/Mail/classes/Mustache/class.ilMailTemplateContextAdapter.php:
```diff
@@ -33,7 +33,8 @@ public function __construct(
         protected array $contexts,
         protected array $context_parameter,
-        protected ?ilObjUser $recipient = null
+        protected ?ilObjUser $recipient = null,
```
Are you sure we need this? I am asking because your colleague Stefan removed this parameter after discussion with (me and) the community here: https://github.com/ILIAS-eLearning/ILIAS/commit/fbe4f34cb5780d70f22efd05e948f953da53693c

lukastocker replied:
is removed

mjansenDatabay commented on components/ILIAS/Mail/classes/Mustache/class.ilMailTemplateContextAdapter.php:
```diff
@@ -64,7 +65,7 @@ static function ($placeholder): string {
     {
         foreach ($this->contexts as $context) {
-            $ret = $context->resolvePlaceholder($name, $this->context_parameter, $this->recipient);
+            $ret = $context->resolvePlaceholder($name, $this->context_parameter, $this->recipient, $this->html_markup);
```
See above

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormGUI.php:
```diff
@@ -55,6 +60,16 @@ class ilMailFormGUI
+    protected ArrayBasedRequestWrapper $query;
+    protected ilMailFormUploadHandlerGUI $upload_handler;
+    protected ilFileDataMail $fdm;
+    protected ILIAS\ResourceStorage\Services $storage;
```
Please make them `private readonly`, if possible.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormGUI.php:
```diff
@@ -55,6 +60,16 @@ class ilMailFormGUI
+    /**
+     * @var \Psr\Http\Message\RequestInterface|\Psr\Http\Message\ServerRequestInterface $request
+     */
+    protected $request;
```
We could add `\Psr\Http\Message\ServerRequestInterface` IMO.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -50,6 +53,8 @@ class ilMail
     private int $max_recipient_character_length = 998;
     private readonly Conductor $legal_documents;
 
+    protected ILIAS\Refinery\Factory $refinery;
```
Can we make this `private readonly`?

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -23,11 +23,14 @@
+    use FileDataRCHandling;
+
+    public const ILIAS_HOST = 'ilias';
+    public const PROP_CONTEXT_SUBJECT_PREFIX = 'subject_prefix';
```
Please re-add the PHP type here.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -87,6 +93,7 @@ public function __construct(
         $this->usr_id_by_login_callable = $usr_id_by_login_callable ?? (static fn(string $login): int => (int) ilObjUser::_lookupId($login));
         $this->auto_responder_service = $auto_responder_service ?? $DIC->mail()->autoresponder();
         $this->user_id = $a_user_id;
+        $this->mail_obj_ref_id = $this->mail_obj_ref_id;
```
Why do we assign a property to itself here?

lukastocker replied:
shouldn't be there - i just removed it

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -302,13 +310,11 @@ public function markRead(array $mail_ids): void
-        $types[] = 'integer';
-        $values[] = 'read';
-        $values[] = $this->user_id;
+        array_push($types, 'text', 'integer');
```
We could revert all changes in method `markRead`.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -323,13 +329,11 @@ public function markUnread(array $mail_ids): void
-        $types[] = 'integer';
-        $values[] = 'unread';
-        $values[] = $this->user_id;
+        array_push($types, 'text', 'integer');
```
We could revert all changes in method `markUnread`.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -345,7 +349,7 @@ public function moveMailsToFolder(array $mail_ids, int $folder_id): bool
         $mail_ids = array_filter(array_map(intval(...), $mail_ids));
 
-        if ([] === $mail_ids) {
+        if (count($mailIds) === 0) {
```
We could revert all changes in method `moveMailsToFolder`.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -385,17 +385,68 @@ public function deleteMails(array $mail_ids): void
+            $row['attachments'] = $id;
+        } else {
+            $row['attachments'] = NULL;
+        }
```
Can't we change this to:

```suggestion
        if (isset($row['attachments']) && is_string($row['attachments']) && $row['attachments'] !== '') {
            $row['attachments'] = new ResourceCollectionIdentification($row['attachments']);
        } else {
            $row['attachments'] = null;
        }
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -385,17 +385,68 @@ public function deleteMails(array $mail_ids): void
         if (isset($row['attachments'])) {
-            $unserialized = unserialize(stripslashes((string) $row['attachments']), ['allowed_classes' => false]);
-            $row['attachments'] = is_array($unserialized) ? $unserialized : [];
+            $unserialized = unserialize(stripslashes($row['attachments']), ['allowed_classes' => false]);
```
`$unserialized` is not used at all.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -385,17 +385,68 @@ public function deleteMails(array $mail_ids): void
-            $unserialized = unserialize(stripslashes((string) $row['attachments']), ['allowed_classes' => false]);
-            $row['attachments'] = is_array($unserialized) ? $unserialized : [];
+            $unserialized = unserialize(stripslashes($row['attachments']), ['allowed_classes' => false]);
+            $row['attachments'] = (string) $row['attachments'] ?? NULL;
```
My static analysis tool says: The operation results to '(string) $row['attachments']', please add missing parentheses.

And: Please use a lowercase `null`.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -506,9 +557,10 @@ private function sendInternalMail(
 
         if ($use_placeholders) {
-            $message = $this->replacePlaceholders($message, $usr_id);
+            $message = $this->replacePlaceholders($message, $usrId);
```
`$usrId` does not exist.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -506,9 +557,10 @@ private function sendInternalMail(
         }
-        $message = str_ireplace(['<br />', '<br>', '<br/>'], "\n", $message);
+        $message = $this->formatLinebreakMessage($this->refinery->string()->markdown()->toHTML()->transform($message) ?? '');
+        $message = str_ireplace(["<br />", "<br>", "<br/>"], "\n", $message);
```
We should not change the format:

```suggestion
        $message = str_ireplace(['<br />', '<br>', '<br/>'], "\n", $message);
```

But: Do we really still need to replace the HTML linebreaks, if we transform `$message` with `$this->refinery->string()->markdown()->toHTML()->transform(...)`?

lukastocker replied:
is now removed

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -555,7 +607,8 @@ private function sendInternalMail(
         string $message,
-        int $usr_id = 0
+        int $usrId = 0,
+        bool $replace_empty_placeholders = true
```
As mentioned in other classes: Do we really need the `$replace_empty_placeholders` logic, since it has been removed some time ago by Stefan Hecken in a PR.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -569,7 +622,8 @@ private function replacePlaceholders(
                 $user,
-                $this->context_parameters
+                $this->context_parameters,
+                $replace_empty_placeholders
```
See above

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -555,7 +607,8 @@ private function sendInternalMail(
     private function replacePlaceholders(
         string $message,
-        int $usr_id = 0
+        int $usrId = 0,
```
This renaming does not work, since the using code below expects `$usr_id`.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -588,112 +642,128 @@ private function replacePlaceholdersEmpty(string $message): string
-        } else {
-            $this->sendMailWithoutReplacedPlaceholder($mail_data, $to_usr_ids, $cc_bcc_recipients);
-        }
+    private function distributeMail(
```
This change (method: `distributeMail`) looks unintended, since it removes the "Autoresponder" feature and re-introduces primitive types being passed instead of `MailDeliveryData`.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -588,112 +642,128 @@ private function replacePlaceholdersEmpty(string $message): string
-        MailDeliveryData $mail_data,
-        array $recipients,
-        string $message
+        string $to,
```
This change (method: `sendChanneledMails`) looks unintended, since it re-introduces primitive types being passed instead of `MailDeliveryData`.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -768,65 +840,93 @@ private function sendChanneledMails(
-        array $attachments,
         string $message,
-        array $usr_id_to_external_email_addresses_map
+        array $attachments,
```
Please only change the relevant code since it is tough to understand the actual code.

And:
`$firstAddresses` != `$first_addresses`

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -986,12 +1089,12 @@ public function enqueue(
 
         $errors = $this->checkMail($a_rcp_to, $a_rcp_cc, $a_rcp_bcc, $a_m_subject);
-        if ($errors !== []) {
+        if (count($errors) > 0) {
```
We should leave this as it is.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -986,12 +1089,12 @@ public function enqueue(
 
         $errors = $this->validateRecipients($a_rcp_to, $a_rcp_cc, $a_rcp_bcc);
-        if ($errors !== []) {
+        if (count($errors) > 0) {
```
We should leave this as it is.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -1013,7 +1116,7 @@ public function enqueue(
 
         if (ilContext::getType() === ilContext::CONTEXT_CRON) {
-            $mail_data = new MailDeliveryData(
+            return $this->sendMail(
```
Why do we re-introduce single primitive types instead of using `ilMailDeliveryData` here?

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -1073,58 +1175,75 @@ public function enqueue(
      */
     public function sendMail(
-        MailDeliveryData $mail_data
+        string $to,
```
Why do we re-introduce single primitive types instead of using `ilMailDeliveryData` here?

It seems there are a lot of unintended changes below. Please revert them.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -1348,8 +1467,8 @@ public static function _getAutoGeneratedMessageString(?ilLanguage $lang = null):
-        global $DIC;
-        $sender_factory = $DIC->mail()->mime()->senderFactory();
+        /** @var ilMailMimeSenderFactory $senderFactory */
+        $senderFactory = $GLOBALS["DIC"]["mail.mime.sender.factory"];
```
This does not work.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -1247,15 +1366,15 @@ private function sendMimeMail(
             $this->table_mail_saved,
             [
-                'attachments' => ['clob', serialize($attachments)],
+                'attachments' => ['clob', $attachments],
```
Here we are use `clob`, above we are using `text` (see: `persistToStage`). This should be streamlined.

lukastocker replied:
Thanks for the hint. I just looked it up in the database. We used varchar and not longtext. Therefore I will change this here to "text".

mjansenDatabay replied:
It should be text, because you migrate the field type from `clob` (LONGTEXT) to `text` (VARCHAR) in the database update objecte in the mail setup agent.

Migrating this to `text` is exactly what we discussed some time ago, so it is perfectly fine. Just use `text` everywhere the `attachment` field is touched for when manipulating the `mail_saved` table.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -1247,15 +1366,15 @@ private function sendMimeMail(
     {
+        if (!is_null($attachments)) {
+            $attachments = $attachments->serialize();
+        }
```
```suggestion
        }

```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -1218,23 +1337,23 @@ private function sendMimeMail(
         }
 
-        if ($bcc !== '') {
+        if ($bcc) {
```
Please revert this.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -1218,23 +1337,23 @@ private function sendMimeMail(
-        if ($cc !== '') {
+        $mailer->Body($this->refinery->string()->markdown()->toHTML()->transform($message) ?? '');
+
+        if ($cc) {
```
Please revert this.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -1133,11 +1252,11 @@ public function sendMail(
-            $random = new Random\Randomizer();
-            if ($random->getInt(0, 50) === 2) {
+            $random = new ilRandom();
+            if ($random->int(0, 50) === 2) {
```
Please revert this.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -1218,23 +1337,23 @@ private function sendMimeMail(
             $subject,
             true,
-            (string) ($this->context_parameters[self::PROP_CONTEXT_SUBJECT_PREFIX] ?? '')
+            (string) ($this->contextParameters[self::PROP_CONTEXT_SUBJECT_PREFIX] ?? '')
```
Please revert this.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFolderGUI.php:
```diff
@@ -899,15 +899,16 @@ protected function showMail(): void
             $radiog = new ilRadioGroupInputGUI('', 'filename');
-            foreach ($mail_data['attachments'] as $file) {
-                $radiog->addOption(new ilRadioOption($file, md5((string) $file)));
+            $mailData['attachments'] = unserialize($mailData['attachments']);
```
Please use the `allowed_classes` option to address security concerns.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormGUI.php:
```diff
@@ -75,7 +90,14 @@ public function __construct(
         $this->mbox = new ilMailbox($this->user->getId());
         $this->purifier = $body_purifier ?? new ilMailBodyPurifier();
-        $this->ui_factory = $DIC->ui()->factory();
+        $this->ui_factory = $ui_factory ?? $DIC->ui()->factory();
```
This will raise an error.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormGUI.php:
```diff
@@ -75,7 +90,14 @@ public function __construct(
-        $this->ui_factory = $DIC->ui()->factory();
+        $this->ui_factory = $ui_factory ?? $DIC->ui()->factory();
+        $this->request = $DIC->http()->request();
+        $this->ui_renderer = $ui_renderer ?? $DIC->ui()->renderer();
```
This will raise an error.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormGUI.php:
```diff
@@ -1033,4 +817,277 @@ protected function showSubmissionErrors(array $errors): void
+        ];
+    }
+
+    protected function addToolbarButtons()
```
```suggestion
    protected function addToolbarButtons(): void
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormGUI.php:
```diff
@@ -1033,4 +817,277 @@ protected function showSubmissionErrors(array $errors): void
+
+                $templates = $this->template_service->loadTemplatesForContextId($context->getId());
+                if (count($templates) > 0) {
+                    $options = array();
```
```suggestion
                    $options = [];
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormGUI.php:
```diff
@@ -515,11 +509,13 @@ public function showForm(): void
-        $mail_data['rcp_cc'] = '';
-        $mail_data['rcp_bcc'] = '';
-        $mail_data['attachments'] = [];
+        $mailData = [];
```
Please have a look at `$mail_data` vs. `$mailData`. It should all be `$mail_data`.

mjansenDatabay reviewed (changes requested):
Hi @lstocker,

thanks again for your work. I left some inline remarks.

Some general aspects:

* There are changes in different components which I cannot approve due to missing authorities (Course, Exercise, Scorm. Session, Survey, StudyProgramme, User, UI). This is not a problem, but I cannot merge this PR later if the others don't approve it as well.
* I currently cannot test the changes. When requesting ILIAS, I immediately receive the following issue (there are a lot more, see my inline comments):

```bash
Whoops\Exception\ErrorException thrown with message "Undefined variable $sender_factory"

Stacktrace:
#16 Whoops\Exception\ErrorException in /var/www/ilias/trunk/components/ILIAS/Mail/classes/class.ilMail.php:1473
#15 ilErrorHandling:handlePreWhoops in /var/www/ilias/trunk/components/ILIAS/Mail/classes/class.ilMail.php:1473
#14 ilMail:_getIliasMailerName in /var/www/ilias/trunk/components/ILIAS/Mail/classes/class.ilPDMailBlockGUI.php:152
#13 ilPDMailBlockGUI:getListItemForData in /var/www/ilias/trunk/components/ILIAS/Block/classes/class.ilBlockGUI.php:573
#12 ilBlockGUI:getListItemGroups in /var/www/ilias/trunk/components/ILIAS/Block/classes/class.ilBlockGUI.php:412
#11 ilBlockGUI:getHTML in /var/www/ilias/trunk/components/ILIAS/Mail/classes/class.ilPDMailBlockGUI.php:102
#10 ilPDMailBlockGUI:getHTML in /var/www/ilias/trunk/components/ILIAS/UICore/classes/class.ilCtrl.php:146
#9 ilCtrl:getHTML in /var/www/ilias/trunk/components/ILIAS/Block/classes/class.ilColumnGUI.php:466
#8 ilColumnGUI:showBlocks in /var/www/ilias/trunk/components/ILIAS/Block/classes/class.ilColumnGUI.php:419
#7 ilColumnGUI:getHTML in /var/www/ilias/trunk/components/ILIAS/UICore/classes/class.ilCtrl.php:146
#6 ilCtrl:getHTML in /var/www/ilias/trunk/components/ILIAS/Dashboard/classes/class.ilDashboardGUI.php:328
#5 ilDashboardGUI:getRightColumnHTML in /var/www/ilias/trunk/components/ILIAS/Dashboard/classes/class.ilDashboardGUI.php:259
#4 ilDashboardGUI:show in /var/www/ilias/trunk/components/ILIAS/Dashboard/classes/class.ilDashboardGUI.php:368
#3 ilDashboardGUI:jumpToSelectedItems in /var/www/ilias/trunk/components/ILIAS/Dashboard/classes/class.ilDashboardGUI.php:231
#2 ilDashboardGUI:executeCommand in /var/www/ilias/trunk/components/ILIAS/UICore/classes/class.ilCtrl.php:119
#1 ilCtrl:forwardCommand in /var/www/ilias/trunk/components/ILIAS/UICore/classes/class.ilCtrl.php:92
#0 ilCtrl:callBaseClass in /var/www/ilias/trunk/public/ilias.php:32
```
  * It would be good if you could test the UI use cases, so that the state of the branch is more or less testable.
* Please apply the PHP-CS-FIXER to all changes PHP files, since the code style seems incorrect here and there (e.g. in `components/ILIAS/Mail/classes/Preview/class.ilMailPreviewGUI.php`).
* Similar to the former pull requests, there are a lot of (IMO) unnecessary (and in my opinion) accidentally applied changes (formatting, renaming of variables, removal of existing behaviour). Please eliminate them.

Best regards,
Michael

mjansenDatabay commented on components/ILIAS/Course/classes/Preview/class.ilCourseMailTemplateTutorContextPreview.php:
```diff
@@ -0,0 +1,94 @@
+
+    /**
+     * @return string
+     */
```
This file does not belong to "my" components, but I suggest:

```suggestion
```

mjansenDatabay commented on components/ILIAS/Course/classes/Preview/class.ilCourseMailTemplateTutorContextPreview.php:
```diff
@@ -0,0 +1,94 @@
+    public const DEFAULT_COURSE_TITLE = "preview_crs_title";
+    public const DEFAULT_COURSE_STATUS = "preview_crs_status";
+    public const DEFAULT_COURSE_MARK = "preview_crs_mark";
+    public const DEFAULT_COURSE_TIME_SPENT = "3671";
```
This file does not belong to "my" components, but I suggest:

```suggestion
    public const string ID = 'crs_context_tutor_manual_preview';
    public const string DEFAULT_COURSE_TITLE = "preview_crs_title";
    public const string DEFAULT_COURSE_STATUS = "preview_crs_status";
    public const string DEFAULT_COURSE_MARK = "preview_crs_mark";
    public const string DEFAULT_COURSE_TIME_SPENT = "3671";
```

mjansenDatabay commented on components/ILIAS/Course/classes/Preview/class.ilCourseMailTemplateTutorContextPreview.php:
```diff
@@ -0,0 +1,94 @@
+
+    /**
+     * {@inheritdoc}
+     */
```
This file does not belong to "my" components, but I suggest:

```suggestion
```

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,141 @@
+trait FileDataRCHandling
+{
+    /**
+     * @param array                       $path_to_files
```
Please specificy the "Shape" of the array according to https://phpstan.org/writing-php-code/phpdoc-types#array-shapes / https://phpstan.org/writing-php-code/phpdoc-types#lists / https://phpstan.org/writing-php-code/phpdoc-types#general-arrays

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,141 @@
+    /**
+     * @param array                       $path_to_files
+     * @param ilMailAttachmentStakeholder $stakeholder
+     * @return ResourceCollection
```
```suggestion
```

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,141 @@
+
+    /**
+     * @param array $mail_data
+     * @return array
```
Please specificy the "Shape" of the array according to https://phpstan.org/writing-php-code/phpdoc-types#array-shapes / https://phpstan.org/writing-php-code/phpdoc-types#lists / https://phpstan.org/writing-php-code/phpdoc-types#general-arrays

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,141 @@
+
+    /**
+     * @param array $mail_data
+     * @return ResourceCollectionIdentification|null
```
```suggestion
```

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    /**
+     * @param array $mail_data
```
Please specify the "Shape" of the array according to https://phpstan.org/writing-php-code/phpdoc-types#array-shapes / https://phpstan.org/writing-php-code/phpdoc-types#lists / https://phpstan.org/writing-php-code/phpdoc-types#general-arrays

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    /**
+     * @param ResourceCollectionIdentification $identification
```
```suggestion
```

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,141 @@
+
+    /**
+     * @param ResourceCollectionIdentification $identification
+     * @return array
```
Please specify the "Shape" of the array according to https://phpstan.org/writing-php-code/phpdoc-types#array-shapes / https://phpstan.org/writing-php-code/phpdoc-types#lists / https://phpstan.org/writing-php-code/phpdoc-types#general-arrays

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,141 @@
+
+    /**
+     * @param array $attachments
+     * @return array
```
Please specify the "Shape" of the array according to https://phpstan.org/writing-php-code/phpdoc-types#array-shapes / https://phpstan.org/writing-php-code/phpdoc-types#lists / https://phpstan.org/writing-php-code/phpdoc-types#general-arrays

mjansenDatabay commented on components/ILIAS/Mail/classes/Object/class.ilMailValueObject.php:
```diff
@@ -28,20 +28,29 @@ class ilMailValueObject
-        private readonly string $recipients_bcc,
-        private readonly string $subject,
-        private readonly string $body,
+        private string $from,
```
These changes in `ilMailValueObject` should be reverted IMO.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormUploadHandlerGUI.php:
```diff
@@ -0,0 +1,119 @@
+
+        return $infos;
+    }
+    public function getInlineConsumer(string $identifier): InlineConsumer
```
```suggestion

    public function getInlineConsumer(string $identifier): InlineConsumer
```

mjansenDatabay commented on components/ILIAS/Mail/tests/ilMailTest.php:
```diff
@@ -634,7 +634,10 @@ private function createAndExpectDatabaseCall(int $some_mail_id, array $row_data)
     private function create(int $ref_id = 234, int $usr_id = 123): ilMail
     {
-        return new ilMail(
+        $refinery = $this->getMockBuilder('\ILIAS\Refinery\Factory')->disableOriginalConstructor()->getMock();
```
```suggestion
        $refinery = $this->getMockBuilder(\ILIAS\Refinery\Factory::class)->disableOriginalConstructor()->getMock();
```

mjansenDatabay commented on components/ILIAS/Mail/tests/ilMailMimeTest.php:
```diff
@@ -75,6 +78,9 @@ public function testMimMailDelegatesEmailDeliveryToDefaultTransport(): void
         ])->getMock();
         $this->setGlobalVariable('ilSetting', $settings);
 
+        $refinery = $this->getMockBuilder('\ILIAS\Refinery\Factory')->disableOriginalConstructor()->getMock();
```
```suggestion
        $refinery = $this->getMockBuilder(\ILIAS\Refinery\Factory::class)->disableOriginalConstructor()->getMock();
```

mjansenDatabay commented on components/ILIAS/Mail/tests/ilMailMimeTest.php:
```diff
@@ -53,6 +53,9 @@ public function testMimMailDelegatesEmailDeliveryToThePassedTransporter(): void
         ])->getMock();
         $this->setGlobalVariable('ilSetting', $settings);
 
+        $refinery = $this->getMockBuilder('\ILIAS\Refinery\Factory')->disableOriginalConstructor()->getMock();
```
```suggestion
        $refinery = $this->getMockBuilder(\ILIAS\Refinery\Factory::class)->disableOriginalConstructor()->getMock();
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormUploadHandlerGUI.php:
```diff
@@ -0,0 +1,119 @@
+        return new BasicFileInfoResult($this->getFileIdentifierParameterName(), $identifier, $r->getTitle(), $r->getSize(), $r->getMimeType());
+    }
+
+    public function getInfoForExistingFiles(array $file_ids): array
```
Please specify the "Shape" of the array according to https://phpstan.org/writing-php-code/phpdoc-types#array-shapes / https://phpstan.org/writing-php-code/phpdoc-types#lists / https://phpstan.org/writing-php-code/phpdoc-types#general-arrays

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilFileDataMail.php:
```diff
@@ -364,7 +363,7 @@ public function saveFile(int $a_mail_id, string $a_attachment): bool
     public function checkFilesExist(array $a_files): bool
     {
-        if ($a_files !== []) {
+        if ($a_files) {
```
IMO this can also be reverted.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilAccountMail.php:
```diff
@@ -188,8 +192,10 @@ public function send(): bool
-        $settings = $DIC->settings();
-        $mustache_factory = $DIC->mail()->mustacheFactory();
+        $tree = $DIC->repositoryTree();
+        $ilSetting = $DIC->settings();
```
As far as I can see, these two dependencies are already assigned in `__construct`.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilAccountMail.php:
```diff
@@ -215,7 +221,7 @@ public function replacePlaceholders(string $a_string, ilObjUser $a_user, NewAcco
         $replacements['ILIAS_URL'] = ILIAS_HTTP_PATH . '/login.php?client_id=' . CLIENT_ID;
         $replacements['CLIENT_NAME'] = CLIENT_NAME;
-        $replacements['ADMIN_MAIL'] = $settings->get('admin_email');
+        $replacements['ADMIN_MAIL'] = $ilSetting->get('admin_email');
```
IMO we could either revert this or use `$this->settings`.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilAccountMail.php:
```diff
@@ -231,9 +237,12 @@ public function replacePlaceholders(string $a_string, ilObjUser $a_user, NewAcco
+        ) {
+            $target = $this->http->wrapper()->query()->retrieve('target', $this->refinery->kindlyTo()->string());
+            $tarr = explode("_", $target);
+            if ($tree->isInTree((int) $tarr[1])) {
```
IMO this can be reverted.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilAccountMail.php:
```diff
@@ -120,7 +124,7 @@ public function send(): bool
         }
 
-        if ($user->getEmail() === '') {
+        if (!$user->getEmail()) {
```
IMO this can be reveted.

mjansenDatabay commented on components/ILIAS/Mail/classes/BackgroundTask/class.ilMailDeliveryJob.php:
```diff
@@ -58,6 +58,7 @@ public function run(array $input, Observer $observer): Value
             }
         }
 
+
```
If we remove this, the whole file won't be listed as changed anymore.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -336,6 +343,7 @@ public function markUnread(array $mail_ids): void
         $this->db->manipulateF($query, $types, $values);
     }
 
+
```
This line can be reverted.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -760,7 +768,7 @@ private function sendChanneledMails(
             );
 
-            if ($mail_data->getAttachments() !== []) {
+            if (count($mail_data->getAttachments()) > 0) {
```
This line can be reverted.

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,141 @@
+     * @param array $mail_data
+     * @return array
+     */
+    public function FilesFromLegacyToIRSS(array $mail_data): array
```
We should rename the method to:

```suggestion
    public function filesFromLegacyToIRSS(array $mail_data): array
```

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,141 @@
+     * @param array $mail_data
+     * @return ResourceCollectionIdentification|null
+     */
+    public function getIDforCollection(array $mail_data): ?ResourceCollectionIdentification
```
We should rename the method to `getIDforCollection` (with a lowercase `d`):

```suggestion
    public function getIdforCollection(array $mail_data): ?ResourceCollectionIdentification
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -51,6 +54,8 @@ class ilMail
     private int $max_recipient_character_length = 998;
     private readonly Conductor $legal_documents;
 
+    private readonly ILIAS\Refinery\Factory $refinery;
```
```suggestion
    private readonly ILIAS\Refinery\Factory $refinery;
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -392,11 +400,10 @@ private function fetchMailData(?array $row): ?array
+            $row['attachments'] = new ResourceCollectionIdentification($row['attachments']);
         } else {
-            $row['attachments'] = [];
+            $row['attachments'] = null;
```
IMO this could be an issue if this method is called from a context, where the mail record was read from database table `mail`, and not `mail_saved`.

In `components/ILIAS/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php`, you only migrated database table `mail_saved`.

So IMO we'll have to handle three different data situations:
- `NULL`
- A `rcid` UUID (I guess for "Drafts")
- For sent mails: A PHP-serialized array of file system paths (array of `string`-elements).

lukastocker replied:
Like discussed in the call recently I will adjust this part to unserialize content if it's saved as serialized string in the database.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -507,17 +514,18 @@ private function sendInternalMail(
         $usr_id = $usr_id ?: $this->user_id;
 
         if ($use_placeholders) {
+            $subject = $this->replacePlaceholders($subject, $usr_id);
```
Do we need this change? This would implicitly introduce placeholders for the subject line, which is IMO currently not stated explicitly as a feature for ILIAS 11.

lukastocker replied:
We do not need this - I will remove this. Thanks for the hint!

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -507,17 +514,18 @@ private function sendInternalMail(
             'folder_id' => ['integer', $folder_id],
             'sender_id' => ['integer', $sender_usr_id],
-            'attachments' => ['clob', serialize($attachments)],
+            'attachments' => ['text', serialize($attachments)],
```
IMO this could be an issue.

In `components/ILIAS/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php`, you only migrated database table `mail_saved`.

So records in table `mail` should be still of type `clob` IMO.

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMail.php:
```diff
@@ -1262,15 +1272,16 @@ private function sendMimeMail(
             $this->table_mail_saved,
             [
-                'attachments' => ['clob', serialize($attachments)],
+                'attachments' => ['clob', $attachments],
```
All database queries and manipulations for table: `mail_saved` / field: `attachments` must use type `text ` IMO, according to the migration in `components/ILIAS/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php`.

```suggestion
                'attachments' => ['text', $attachments],
```

We discussed this above and you responded with a :+1: , but it seems to be still the wrong type :smile: .

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormGUI.php:
```diff
@@ -198,28 +224,42 @@ public function sendMessage(): void
+            $value['use_placeholders']
+        )
+        ) {
+            $_POST['attachments'] = $files;
```
We MUST not manipulate the super global `$_POST` array.

mjansenDatabay reviewed (changes requested):
Hi @lukastocker,

thanks for your changes, I think we are almost there! I have added some additional inline comments.

Besides the minor/trivial issues, we should double-check the database persistence of attachments in the context of the following tables:

1. `mail`
2. `mail_saved`

You migrated the `attachments` field of the latter to type `text` (`VARCHAR(100)`) with new database update steps in `components/ILIAS/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php`, whereas the field in `mail` is (and IMO should remain) of type `clob` (`LONGTEXT`).  

The `mail` table field can still contain a serialized array of attachment file paths (string).

Please make sure that:

- [ ] The correct field type is always used depending on the table being queried or manipulated.
- [ ] The value of this field is always treated correctly, either `null` or an IRSS ID (`string`) in `mail_saved`, and either `null` or a serialized PHP `string` list/array in `mail`.
- [ ] After verifying type and value handling, you perform a complete GUI workflow:
  - compose a new mail,
  - add attachments to the stage,
  - assign them to the new mail,
  - add recipients (e.g. from mailing lists, courses, or groups) by navigating to the "Contacts" screens (by using the toolbar buttons on top of the "Compose Mail" form),
  - save and re-edit the mail as a **draft**,
  - finally, send it and verify the details of the sent mail in the folder and detail view (of the sender and the recipients).

Additionally, could you please clarify one point:

- [ ] Did you test all encoding/decoding and transformation steps also with existing *legacy* mail records (stored in table `mail`)?  
  I’m referring to records that don’t include any Markdown syntax.  
  With this PR, several transformations are involved, on the one hand: `html_entity_decode` and `$this->refinery->string()->markdown()->toHTML()`. On the other hand, I also see some removed `ilUtil::securePlainString()` calls.

Best regards,  
Michael

lukastocker commented:
Hi @mjansenDatabay ,

as I can't check the boxes in your [comment](https://github.com/ILIAS-eLearning/ILIAS/pull/9423#pullrequestreview-3295041475) I will list the changes pushed last below:

- unnecessary code was reverted/removed
- fetching mail data includes a check for serialized strings and unserializes them. This resulted in some changes for e.g. draft mails too as unserializing was already done during fetching the mail data 
- table mail and mail_saved column type were adjusted (text and clob)
- return values (array <type>) were adjusted accordingly
- PHPUnit tests were adjusted
- the GUI tests you mentioned were performed successfully 
- encoding/decoding and transformation steps with legacy mails were tested successfully
- I pulled the current trunk code and resolved conflicts

Greetings
Luka

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,136 @@
+            $rcid = $this->storage->collection()->id();
+            $collection = $this->storage->collection()->get($rcid);
+            foreach ($path_to_files as $path_to_file) {
+                $path_to_file = str_replace('/var/ilias/data/ilias', '', $path_to_file);
```
This path does not exist on my system ;-).

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,136 @@
+
+    /**
+     * @param array<string, mixed> $mail_data
+     * @return array<mixed, null>
```
I suggest ... 

```suggestion
     * @return list<string>
```

..., since `\ilFileDataMail::getAbsoluteAttachmentPoolPathByFilename` as well as `\ILIAS\ResourceStorage\Identification\AbstractIdentification::serialize` return a string.

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,136 @@
+trait FileDataRCHandling
+{
+    /**
+     * @param array<string> $path_to_files
```
```suggestion
     * @param list<string> $path_to_files
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilAccountMail.php:
```diff
@@ -188,8 +192,8 @@ public function send(): bool
-        $settings = $DIC->settings();
-        $mustache_factory = $DIC->mail()->mustacheFactory();
+        /** @var ilMustacheFactory $mustache_factory */
+        $mustache_factory = $DIC["mail.mustache.factory"];
```
I suggest reverting this because the key `mail.mustache.factory` is only set in `$DIC` if we can ensure that `\ILIAS\Mail\Service\MailService::__construct` was called once.

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    /**
+     * @return array<string>
```
```suggestion
     * @return list<string>
```

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,136 @@
+        foreach ($all_ids as $id) {
+            $files[] = $id->serialize();
+        }
+        return $files;
```
```suggestion

        return $files;
```

mjansenDatabay commented on components/ILIAS/Mail/classes/Attachments/FileDataRCHandling.php:
```diff
@@ -0,0 +1,136 @@
+
+    /**
+     * @param array<string, mixed> $attachments
+     * @return array<string>
```
```suggestion
     * @return list<string>
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormUploadHandlerGUI.php:
```diff
@@ -0,0 +1,124 @@
+            return new BasicHandlerResult($this->getFileIdentifierParameterName(), HandlerResultInterface::STATUS_OK, $identifier, 'file deleted');
+        } else {
+            return new BasicHandlerResult($this->getFileIdentifierParameterName(), HandlerResultInterface::STATUS_FAILED, $identifier, 'file not found');
+        }
```
```suggestion
        }
        
        return new BasicHandlerResult($this->getFileIdentifierParameterName(), HandlerResultInterface::STATUS_FAILED, $identifier, 'file not found');
```

mjansenDatabay commented on components/ILIAS/Mail/classes/class.ilMailFormUploadHandlerGUI.php:
```diff
@@ -0,0 +1,124 @@
+    /**
+     * @param array<string, mixed> $file_ids
+     * @return array<BasicFileInfoResult>
+     */
```
Sorry, I just saw that the type is already specified in the base class `\ILIAS\UI\Component\Input\Field\UploadHandler::getInfoForExistingFiles`.

My bad ;-).

```suggestion
```

mjansenDatabay reviewed (approved):
Thanks @lukastocker and @klees .

I am now fine with the "Mail"-related changes.

Now we need the approval from the responsible authorities for:

- Course (@smeyer-ilias)
- UI (@klees / @thibsy)
- StudyProgramme (@klees)
- User (@kergomard)

mjansenDatabay commented:
@lukastocker I noticed one issue when using your branch. When clicking on a placeholder in the mail form, the form gets submitted and validated. This should be changed.

mjansenDatabay committed:
Mail: Implement "User Search" based on public API

mjansenDatabay commented:
I imlemented a "User Search"-based recipient search here https://github.com/mjansenDatabay/ILIAS/tree/tr_mark_final_autocomplete (it will always be the latest commit in this branch, not matter how often I'll rebase the branch).

klees reviewed (approved):
I approve the changes in the Study Programme.

I also informally approve the changes in UI (as I have discussed them with @lukastocker some time ago...), but I would kindly ask @thibsy for a formal approval.

Also, we should take this to the JF...

lukastocker commented:
Hi @mjansenDatabay ,
the placeholder issue is now resolved.
Also html will be rendered correctly for e.g. print view etc. and the required option was removed from the mail content area.
In addition I cherry-picked your user search commit to this branch

thibsy commented:
FYI: I will try to review the UI changes today/tomorrow and implement some (or all) of the requested changes inside another PR myself, which I will open against @lukastocker's base repository. This should speed up things so we can get this included in ILIAS 11.

thibsy committed:
[FIX] restore trunk state of `UI/` and `UI_/`.

thibsy committed:
[FEATURE] UI: implement `Field\HasMustacheVariables` minmially invasive.

thibsy committed:
[FEATURE] UI: implement `Form\Standard::withAdditionalFormAction()` minmially invasive and in the right place.

thibsy reviewed (changes requested):
Hi @lukastocker,

Thanks a lot for your contribution to the UI components!

For future reference, please try to create separate PRs for changes in UI – and split up different kinds of features/functionalities. Having changes like this inside large feature commits makes a) reviewing and b) reverting, cherry-picking or simply tracking down changes much harder.

Before you continue reading or act on this review, please incorporate the changes I implemented in https://github.com/lukastocker/ILIAS/pull/7.

Then, for you and the general public as well: if I understand the [mustache specification](https://mustache.github.io/mustache.5.html) correctly, `{{` and `}}` are **delimiters**, using them like `{{}}` creates a **tag** and tags like `{{foo}}` or `{{foo.bar}}` are considered **variables**. Therefore the proper name for this feature is IMO "mustache variables", not placeholders.

Please consider the following suggestions. You do not need to follow them, but please indicate shortly why you prefer to do otherwise:

- [x] `_ui-component_markdown.scss`: can you explain why these changes were necessary? If they are not really required, please remove them.

Please implement the following changes:

- [ ] Interface changes: I have made some interface changes without touching the usages in other components of this PR. Please go ahead an amend them.
- [x] Styling: I have changed the HTML structure of how mustache variables are displayed. Please check if they need styling, and if so, implement it.
- [ ] Example descriptions: I only added the three examples without description. Please go ahead and create them, so they can be used for test cases on Testrail.
- [ ] Unit tests: Please implement some PHP unit tests for the rendering of mustache variable definitions. And maybe one checking the mutators (valid invalid array values).
- [x] Rebase: please rebase this onto latest trunk.

Some personal notes for the Roadmap:

* I noticed that the dropzone and roundtrip components extend from the form container. I was personally responsible for this =) and now admit to my mistake. We should get rid of this entanglement, since i.e. the new `withAdditionalFormAction()` does nothing for roundtrips (rendered without form buttons).
* The JavaScript facility heavily relies on the textarea class. So does the initialisation code. We should probably refactor this, so features like `insertCharactersAroundSelection()` can be used for i.e. the normal text fields as well. This should achieve less or weaker coupling.
* I also noticed the need for some `InputFactory`, which stores different kinds of input fields. I see how Textarea Markdown and Text are at least supporters of mustache variables. Initialising the same current facility for inserting variables requires knowledge about the type so we know from what factory to retrieve an instance. This should be avoided somehow, currently I see this as a solution.

Kind regards,
@thibsy

lukastocker committed:
Mail Markdown: adjustments after cherry-picking changes regarding placeholders and buttons.

lukastocker commented:
Hi @thibsy ,

I just pushed following changes:

- reverted some changes in file.js because they got lost through the last cherry-picks, which resulted into false additional button actions
- implemented changes into the interface
- wrote some PHPUnit tests
- added example descriptions and expected outputs
- rebased trunk with this PR
- styled the mustache variables listing via scss and deleted unnecessary scss regarding markdown preview

Currently there's is a problem with saving and sending attachments. I will look into it on monday.

thibsy commented on components/ILIAS/UI/resources/js/Input/Field/file.js:
```diff
@@ -337,9 +345,9 @@ il.UI.Input = il.UI.Input || {};
 			}
 
 			// abort if the given file is not an allowed file type.
-			if (dropzones[input_id].options.acceptedFiles &&
-				!Dropzone.isValidFile(file, dropzones[input_id].options.acceptedFiles)
-			) {
```
I don't believe this change is necessary, most likely a rebase error. Using `Dropzone.isValidFile()` here specifically addressed a mantis issue regarding mime-types.

thibsy commented on components/ILIAS/UI/resources/js/Input/Field/file.js:
```diff
@@ -416,6 +424,9 @@ il.UI.Input = il.UI.Input || {};
 			if (dropzone.options.form.should_submit === true && current_dropzone >= current_dropzone_count) {
+        if (submit_event.submitter !== null && submit_event.submitter.hasAttribute('formaction')) {
+          dropzone.options.form.action = submit_event.submitter.formAction;
+        }
```
Please get rid of this code and simply call `requestSubmit()` with the `submitter` button. This logic should be handled by this function call.

lukastocker replied:
Not changed. See: https://github.com/ILIAS-eLearning/ILIAS/pull/9423#issuecomment-3451602771

thibsy commented on components/ILIAS/UI/resources/js/Input/Field/file.js:
```diff
@@ -642,6 +653,9 @@ il.UI.Input = il.UI.Input || {};
                 if (total_files === 0) {
+                  if (event.submitter !== null && event.submitter.hasAttribute('formaction')) {
+                    form.action = event.submitter.formAction;
+                  }
```
same here.

lukastocker replied:
Not changed. See: https://github.com/ILIAS-eLearning/ILIAS/pull/9423#issuecomment-3451602771

thibsy commented on components/ILIAS/UI/resources/js/Input/Field/file.js:
```diff
@@ -651,7 +665,10 @@ il.UI.Input = il.UI.Input || {};
+                  if (event.submitter !== null && event.submitter.hasAttribute('formaction')) {
+                    form.action = event.submitter.formAction;
+                  }
+                  form.submit();
```
same here.

lukastocker replied:
Not changed. See: https://github.com/ILIAS-eLearning/ILIAS/pull/9423#issuecomment-3451602771

thibsy commented on components/ILIAS/UI/src/Implementation/Component/Input/Field/Renderer.php:
```diff
@@ -517,21 +517,23 @@ public function renderSelectField(F\Select $component, RendererInterface $defaul
+                        id,
+                        '{$component->getMarkdownRenderer()->getAsyncUrl()}',
+                        '{$component->getMarkdownRenderer()->getParameterName()}'
+                    );
```
Please use the version from trunk here (no `const id =`), this addressed a mantis issue.

thibsy commented on components/ILIAS/UI/src/Implementation/Component/Input/Field/Renderer.php:
```diff
@@ -614,9 +614,56 @@ protected function getPreparedTextareaTemplate(F\Textarea $component): Template
+            function ($val) {
+                $val = htmlentities((string) $val);
+                return str_replace('{{', '&lcub;&lcub;', str_replace('}}', '&rcub;&rcub;', $val));
+            }
```
Please move this into a `$this->mustacheVariableEntities()` method like so future components can reuse this.

thibsy commented on components/ILIAS/UI/resources/js/Input/Field/file.js:
```diff
@@ -294,6 +300,8 @@ il.UI.Input = il.UI.Input || {};
 			// for NoSubmit signals, this can also be an HTMLButtonElement.
 			dropzone.options.form.should_submit = true;
 
+      submit_event = event;
```
please extract the submitter here and rename the variable accordingly. We don't need the entire event.

thibsy reviewed (changes requested):
Hi @lukastocker,

Thx for implementing the remaining changes. There are some things left to do before we can merge this, I mostly added them as inline comments.

In addition to them, please:

- [ ] Examples: be more descriptive with the expected output description of all three examples. This will be the base for our Testrail test cases, so testers need to know a bit more about where to find the output and what it should look like. Assume they don't know what mustache variables are, or what an anchor is etc.

Thx and kind regards,
@thibsy

lukastocker commented:
Hi @thibsy , @mjansenDatabay 

I updated this PR with following fixes:

1. attachments can be saved as draft and send as some function got lost during the cherry pick.
2. placeholders are now rendered correctly if a mail was saved. It already worked for sending mails.
3. adjusted the code according to the comments above for some parts. After some debugging with thibsy we decided to leave the submitter actions in the file.js as they are for now. Comments were added in the file as the file.js should get refactored at some point.
4. adjusted the example output descriptions.
5. rebased PR with the trunk branch to avoid missing code and/or conflicts

I tested the functionality via GUI and ran PHPUnit tests.

Note, that currently there are some issues with the data selection in the data table. This seems to be a general issue as it does not work in the UI docu too. I think we should not fix it in this PR as it is not part of the mail markdown topic.

thibsy reviewed (approved):
Thx @lukastocker for bearing with me. Changes to the UI components LGTM now, I will leave merging to the other authorities though.

I added the JF label because we will need to announce the public interface changes after merge.

lukastocker committed:
Mail Markdown: fix some issues:

Markdown Preview is now rendered in the correct position.
Mail Content is rendered correctly in external mails.

mjansenDatabay commented:
@smeyer-ilias / @kergomard Is there any chance you can review this until ~ 4 PM/16:00 ?

kergomard reviewed (approved):
Changes in User look good to me. Thank you very much!

smeyer-ilias reviewed (approved):
Thanks @lukastocker

matthiaskunkel commented:
Jour Fixe, 10 NOV 2025: Yvonne notified us about changes in the public interface through this PR. Changes were already committed to trunk before release 11 branch has been created and are therefore available in ILIAS 11, too.
